tslearn/early_classification.py

```diff
@@ -0,0 +1,312 @@
+        probas = 1.0 / (
+            np.exp(self.lamb * minimum_distance_to_average)
+            + np.exp(self.lamb * (minimum_distance_to_average - delta))
+            + 1e-6
+        )
+        return probas / probas.sum()
+
+    def _expected_cost(self, X, tau):
+        """
+        From a incomplete series xt, this function compute the expected cost of a prediction made at time "last time of
+        xt + tau"
+
+        Parameters
+        ----------
+        X: vector
+            a time series that may be truncated for which we want to have the cost
+        tau: int
+            gives the future time for which the cost is calculated
+
+        Returns
+        --------
+        float : the computed cost
+        """
+        cost = 0
+        proba_clusters = self.get_cluster_probas(X=X)
+        truncated_t = X.shape[-1]
+        for k in range(0, self.n_clusters):
+            sum_pyhatyck = np.sum(self.pyhatyck_["pyhatycks{0}".format(k)][
+                                truncated_t + tau - self.minimum_time_stamp
+                            ], axis=0)
+            sum_global = np.dot(sum_pyhatyck, self.pyck_[:, k])
+            cost = cost + proba_clusters[:, k] * sum_global
+
+        return cost
+
+    def minimize_integer(self, end_of_time, xt):
+        """
+        We want to minimize a function "funct" according a time series "xt" for a list of integers from 0 to
+        "end_of_time"
+
+        Parameters
+        ----------
+        end_of_time: int
+            the highest integer at which the function to be minimized is calculated
+        function_to_minimize: funct
+            the function of interest
+        xt: vector
+            the time series related to the function "funct"
+
+        Returns
+         ------
+         int : The integer in {0,...,stop} that minimizes "funct"
+         float : the so fat minimum cost
+        """
+        tau_star = 0
+        so_far_minimum_cost = self._expected_cost(xt, tau_star)
+        for tau in range(1, end_of_time):
+            current_cost = self._expected_cost(xt, tau)
+            if current_cost < so_far_minimum_cost:
+                so_far_minimum_cost = current_cost
+                tau_star = tau
+        return tau_star + xt.shape[-1], so_far_minimum_cost
+
+    def _predict_single_series(self, Xi):
+        """
+        This function classifies a single time series xt
+
+        Parameters
+        ----------
+        xt: vector
+            a time series that is probably incomplete but that nonetheless we want to classify
+        Returns
+        -------
+        int: the class which is predicted
+        int : the time of the prediction
+        float : the probability used for computing the cost
+        float : the loss when classifying
+        """
+        length_Xi = Xi.shape[0]
+        Xi = np.reshape(Xi, (1, length_Xi))
+        time_prediction = self.minimum_time_stamp
+        stop_criterion = False
+        while stop_criterion is False:
```

Comment:
I would rather use a for loop (to avoid infinite loops in corner cases) here since you are basically iterating over consecutive possible prediction times, with a `break` instruction if the stopping criterion is met.
Also, it seems that cases where (i) t=max_t and (ii) current_expected_loss <= min future_expected_losses lead to the same actions so they should probably be put together in a single `if` statement to avoid code redundancy.